neat/phenotype/feed_forward.py

```diff
@@ -1,65 +1,56 @@
 import torch
 import torch.nn as nn
+
 import neat.activations as a
-from torch import autograd
+
+
+dtype = torch.float64
+device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
 
 
 class FeedForwardNet(nn.Module):
 
     def __init__(self, genome, config):
         super(FeedForwardNet, self).__init__()
         self.genome = genome
+        self.config = config
         self.units = self.build_units()
         self.lin_modules = nn.ModuleList()
-        self.config = config
         self.activation = a.Activations().get(config.ACTIVATION)
-
-        for unit in self.units:
-            self.lin_modules.append(unit.linear)
+        self.bias_indices = [u.ref_node.id for u in self.units if u.is_bias()]
+        self.input_indices = torch.tensor([u.ref_node.id for u in self.units if u.is_input()], device=device)
+        self.output_indices = torch.tensor([u.ref_node.id for u in self.units if u.is_output()], device=device)
+        self.stacked_units = self.genome.order_units(self.units)
+        self.units_indexes = {u.ref_node.id: k for k, u in enumerate(self.units)}
+        self.units_input_indices = [
+            torch.tensor(self.genome.get_inputs_ids(u.ref_node.id), dtype=torch.long, device=device) for u in self.units
+        ]
+        self.extras_len = len(self.units) - len(self.input_indices)
+        self.build_modules()
 
     def forward(self, x):
-        outputs = dict()
-        input_units = [u for u in self.units if u.ref_node.type == 'input']
-        output_units = [u for u in self.units if u.ref_node.type == 'output']
-        bias_units = [u for u in self.units if u.ref_node.type == 'bias']
-        stacked_units = self.genome.order_units(self.units)
-
-        # Set input values
-        for u in input_units:
-            outputs[u.ref_node.id] = x[0][u.ref_node.id]
-
-        # Set bias value
-        for u in bias_units:
-            outputs[u.ref_node.id] = torch.ones((1, 0)).to(device)[0]
+        inputs = torch.index_select(x, 1, self.input_indices)
+        extras = torch.ones((1, self.extras_len), dtype=dtype, device=device)
+        output_tensor = torch.cat((inputs, extras), 1)
```

Comment:
Seems to cause this error:

```
Traceback (most recent call last):
  File "xor_run.py", line 19, in <module>
    solution, generation = neat.run()
  File "/neat/population.py", line 26, in run
    genome.fitness = max(0, self.Config.fitness_fn(genome))
  File "/neat/experiments/xor/config.py", line 58, in fitness_fn
    pred = phenotype(input)
  File /anaconda3/envs/pytorch-neat/lib/python3.5/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/neat/phenotype/feed_forward.py", line 34, in forward
    output_tensor = torch.cat((inputs, extras), 1)
RuntimeError: Expected object of scalar type Float but got scalar type Double for sequence element 1 in sequence argument at position #1 'tensors'
```

Reply:
@crackcomm I assume you don't get this error. What version of torch and python are you running?

Reply:
python 3.7 and torch 1.0.1

it seems input tensor is missing `dtype = torch.float64`